### PR TITLE
169735612 create contact us mvc

### PR DESCRIPTION
**Pull request by radek-polcode**
#### What's this PR do?
- Create a "Contact Us" feature that saves the contact requests in a database

#### How should this be manually tested?
1.
- Go to http://test.fmadata.com:3000/ in order to properly use hCaptcha
- Click `Contact us` link and try to send new contact request
2. 
- Create admin user (is_admin: true) 
- Login as admin
- Try to see contact requests on `/contact_requests` path

#### What are the relevant tickets?
- https://www.pivotaltracker.com/story/show/169735612

#### Task completed checklist:
- [x] Is there appropriate test coverage?
- [x] Did I take into consideration possible security vulnerabilities?
- [x] Are the controllers secure?
- [x] Was every important matter documented?

#### Things to watch out for
- [N] Does this PR have any migrations?
- [N] Does this PR add new dependencies?

### CODE REVIEW

File: app/controllers/public_controller.rb [hunk 1]
```diff
@@ -21,16 +21,49 @@ def privacy_policy
...
+    else
+      @contact_request = ContactRequest.new
+    end
+    @title = t('public.contact_us.title')
+  end
+
   def saw_cookie_notice
     cookies[:saw_cookie_notice] = true
-    flash[:info] = nil
+    flash.clear
```
Comment (oborze) [inline app/controllers/public_controller.rb]: Just clear the info about the cookies, there could be some other flash alert/notice that will get deleted this way.

File: app/views/layouts/_flash.html.erb [hunk 2]
```diff
@@ -1,8 +1,15 @@
...
-    <div class="alert <%= bootstrap_class_for_flash(type) %> alert-dismissable fade show">
-      <%= msg %>
-      <button class="close" data-dismiss="alert">x</button>
-    </div>
+    <% if flash_message_is_cookie_info(msg) && cookies[:saw_cookie_notice] == nil %>
+      <div class="alert <%= bootstrap_class_for_flash(type) %> alert-dismissable fade show">
+        <%= msg %>
+        <%= link_to('X', saw_cookie_notice_path, method: :post, class: 'float-right') unless cookies[:saw_cookie_notice] %>
+      </div>
+    <% else %>
```
Comment (oborze) [inline app/views/layouts/_flash.html.erb]: There could be both the cookie notice + some other flash alerts/notice so this should be a separate if statement

File: app/controllers/public_controller.rb (hunk 1 above)
Comment (oborze) [inline app/controllers/public_controller.rb]: We should clear just the cookie notice, there might be some other flash messages that'll get fully wiped out this way

File: app/views/layouts/_flash.html.erb (hunk 2 above)
Comment (oborze) [inline app/views/layouts/_flash.html.erb]: This should be a separate if statement - if there's still a cookie notice, other alerts won't get displayed

File: test/models/contact_request_test.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+require 'test_helper'
+
+class ContactRequestTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
+  # end
```
Comment (oborze) [inline test/models/contact_request_test.rb]: Remove this comment

File: test/mailers/contact_request_mailer_test.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
```
Comment (oborze) [inline test/mailers/contact_request_mailer_test.rb]: Remove this file if we're not using it

File: app/views/public/_form_contact_us.html.erb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  <div class='form-group col-md-6'>
+    <%= label_tag(:email_address, nil, class: 'required') %>
+    <%= f.text_field :email_address, class: 'form-control', required: true %>
+  </div>
+  <div class='form-group col-md-6'>
+    <%= label_tag(:telephone) %>
+    <%= f.text_field :telephone, class: 'form-control' %>
+  </div>
+  <div class='form-group col-md-6'>
+    <%= label_tag(:comments) %>
```
Comment (oborze) [inline app/views/public/_form_contact_us.html.erb]: I'd call this "message" instead

File: test/controllers/public_controller_test.rb [hunk 6]
```diff
@@ -25,4 +41,22 @@ class PublicControllerTest < ActionDispatch::IntegrationTest
...
+      post contact_us_url, params: @contact_us_params
+    end
+    assert_response :success
+  end
+
+  test 'should send email to admins' do
+    assert_difference('ActionMailer::Base.deliveries.count') do
+      post contact_us_url, params: @contact_us_params
+    end
+  end
```
Comment (oborze) [inline test/controllers/public_controller_test.rb:98]: Please add tests for failed model validations

File: app/views/layouts/_flash.html.erb [hunk 7]
```diff
@@ -2,7 +2,11 @@
   <% flash.each do |type, msg| %>
     <div class="alert <%= bootstrap_class_for_flash(type) %> alert-dismissable fade show">
       <%= msg %>
-      <button class="close" data-dismiss="alert">x</button>
+      <% if flash_message_is_cookie_info(msg) && cookies[:saw_cookie_notice] == nil %>
+        <%= link_to('X', saw_cookie_notice_path, method: :post, class: 'float-right') %>
+      <% else %>
+        <button class="close" data-dismiss="alert">x</button>
+      <% end %>
```
Comment (oborze) [inline app/views/layouts/_flash.html.erb:9]: When both an alert and a cookie flash is present, the "close" button won't be displayed. We need to display both the cookie info + alert message under each other.

File: test/controllers/public_controller_test.rb (hunk 6 above)
Comment (oborze) [reply test/controllers/public_controller_test.rb:98]: I meant the controller tests for failed model validation - invalid contact params, e.g. missing email, but yeah, my comment was not clear enough

File: app/views/layouts/_flash.html.erb (hunk 7 above)
Comment (radek-polcode) [reply app/views/layouts/_flash.html.erb]: @oborze I think you're wrong here, please check once again. I have cookie alert and ie. info about sign in in two separate boxes, both of them have close button.

File: test/controllers/public_controller_test.rb (hunk 6 above)
Comment (radek-polcode) [reply test/controllers/public_controller_test.rb]: Added.

File: app/views/layouts/_flash.html.erb (hunk 7 above)
Comment (oborze) [reply app/views/layouts/_flash.html.erb:9]: Yup, sorry!

File: app/controllers/contact_requests_controller.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+class ContactRequestsController < ApplicationController
+  before_action :authenticate_user!, only: [:index]
+
+  def index
+    if current_user.try(:is_admin?)
```
Comment (pawjer) [inline app/controllers/contact_requests_controller.rb]: You can use there safe navigation operator `&.`.

File: app/controllers/public_controller.rb [hunk 9]
```diff
@@ -21,16 +21,49 @@ def privacy_policy
   end
 
   def check_if_saw_cookie_notice
-    flash[:info] = "This site uses cookies. See our <a href='/cookie-policy'>Cookie Policy</a> for details.".html_safe unless cookies[:saw_cookie_notice]
+    unless cookies[:saw_cookie_notice]
+      flash[:info] = "This site uses cookies. See our <a href='/cookie-policy'>Cookie Policy</a> for details.".html_safe
```
Comment (pawjer) [inline app/controllers/public_controller.rb]: Don't we need to use i18n here as well?

File: app/helpers/application_helper.rb [hunk 10]
```diff
@@ -13,4 +13,8 @@ def bootstrap_class_for_flash(flash_type)
       flash_type.to_s
     end
   end
+
+  def flash_message_is_cookie_info(msg)
+    msg == "This site uses cookies. See our <a href='/cookie-policy'>Cookie Policy</a> for details."
```
Comment (pawjer) [inline app/helpers/application_helper.rb]: i18n?

File: app/views/layouts/application.html.erb [hunk 11]
```diff
@@ -83,6 +83,14 @@
           <li class="nav-item">
             <%= link_to 'Terms of Use', terms_of_use_path, class: 'nav-link' %>
```
Comment (pawjer) [inline app/views/layouts/application.html.erb]: i18n?

File: app/views/public/_form_contact_us.html.erb [hunk 12]
```diff
@@ -0,0 +1,36 @@
+<%= form_with model: @contact_request, 
+              class: 'form',
+              method: 'POST',
+              local: true,
+              url: '/contact-us' do |f| %>
+  <% if @contact_request.errors.any? %>
+    <div class="form-errors">
+        <h4><%= pluralize(@contact_request.errors.count, "error") %> stopped you from sending contact request:</h4>
```
Comment (pawjer) [inline app/views/public/_form_contact_us.html.erb]: i18n?

**Pull request merged**